### ISSUE TITLE
Reading replies with VoiceOver is extremely inefficient. This is an example and a proposed technical solution.

### ISSUE THREAD

**Issue by Simon818** [TelegramMessenger/Telegram-iOS#504]
At some point recently, Telegram changed the accessibility label on message replies so that VoiceOver users would know which message was being replied to. Now, when you navigate to a message that is a reply to another message, VoiceOver will first read out the entire original message, and then read the reply which is currently in focus. This forces me to listen to the entire (possibly long) original message in order to hear the reply--the message I'm actually trying to read. This dramatically slows down reading of messages, because we have to keep listening to the same messages over and over to get to the reply to the message. Most of the time, hearing "reply to <person's name>" is enough to figure out which message is being replied to, so this is very redundant.

My solution: Change the accessibility label so that it tells us who the message is replying to, but does not read the original message. Add the original message to the accessibility *hint* instead of the accessibility label. This hint will be spoken after a delay depending on settings, and can also be spoken with a three-finger tap. This gives the user the flexibility to hear the original message when they want to, which is probably less than 50% of the time. They should only need to look at the original message after hearing the reply and realizing they don't have enough context.

If this is confusing, read the following imaginary group chat, written as VoiceOver would speak it. Keep in mind that  VoiceOver user cannot skim the message to get to the actual reply; we can only navigate from control to control--in this case, from message to message. So the text we hear first should be the most important text. maybe it would help to actually read this verbatim out loud.

> Message from Alex: I'm inviting everyone over for surprise dinner tomorrow at 5. You're all going to be my guinea pigs as I learn how to use the Instant Pot, but I'm pretty confident it's going to be delicious! - March 14th at 3:30 PM

> Reply to message from Alex: Reply to message from Alex: I'm inviting everyone over for surprise dinner tomorrow at 5. You're all going to be my guinea pigs as I learn how to use the Instant Pot, but I'm pretty confident it's going to be delicious! - March 14th at 3:30 PM. Beckett: Sounds great, I'll be there! - March 14th at 3:31 PM

> Reply to message from Alex: Reply to message from Alex: I'm inviting everyone over for surprise dinner tomorrow at 5. You're all going to be my guinea pigs as I learn how to use the Instant Pot, but I'm pretty confident it's going to be delicious! - March 14th at 3:30 PM. Charlie: Me too! - March 14th at 3:34 PM

> Reply to message from Alex: Reply to message from Alex: I'm inviting everyone over for surprise dinner tomorrow at 5. You're all going to be my guinea pigs as I learn how to use the Instant Pot, but I'm pretty confident it's going to be delicious! - March 14th at 3:30 PM. Your message: Me too! Can I have a hint? - March 14th at 4:02 PM

> Reply to your message: Simon: Me too! Can I have a hint? - March 14th at 4:02 PM. Alex: It will involve chicken. That's all I'm willing to say. - March 14th at 4:17 PM

I deliberately kept the replies short, to illustrate the fact that one would need to listen through these ridiculously long messages to hear a tiny reply. A sighted person could skim this conversation in about 10 seconds--and so could I, if I only heard the replies. To accurately read this conversation with VoiceOver, I would have to hear Alex's original message four times, and it would probably take close to a minute with average speech rate settings. Also, the double "reply to message from" is not a typo; VoiceOver actually says this twice for some reason. The punctuation and timestamps might have been off, but the way I wrote these replies was accurate to how VoiceOver actually reads them in one of my groups.

Now, here is the way this conversation would look with my solution implemented:

> Message from Alex: I'm inviting everyone over for surprise dinner tomorrow at 5. You're all going to be my guinea pigs as I learn how to use the Instant Pot, but I'm pretty confident it's going to be delicious! - March 14th at 3:30 PM

> Beckett, in reply to Alex: Sounds great, I'll be there! - March 14th at 3:31 PM

(At this point, if for some strange reason I didn't know which message Beckett was replying to, I could three-finger tap and have it read to me. Since it's pretty obvious to anyone reading the conversation, they can safely move on without having to hear that message again.)

> Charlie, in reply to Alex: Me too! - March 14th at 3:34 PM

> Your reply to Alex: Me too! Can I have a hint? - March 14th at 4:02 PM

> Alex, in reply to you: It will involve chicken. That's all I'm willing to say. - March 14th at 4:17 PM
